src/displayobjects/rectangle.js

```diff
@@ -1,68 +1,219 @@
+					}
+
+					// Do color if stroke width is specified
+					if (this.strokeWidth > 0) {
+
+						// Set styles
+						canvas.lineWidth = this.strokeWidth;
+						canvas.strokeStyle = this.strokeColor;
+
+						// Set stroke outside the box
+						if (this.strokePosition === "outside") {
+							canvas.strokeRect(x - this.strokeWidth / 2, y - this.strokeWidth / 2, this.width + this.strokeWidth, this.height + this.strokeWidth);
+						}
+
+						// Set stroke on the edge of the box (half of the stroke outside, half inside)
+						else if (this.strokePosition === "center") {
+							canvas.strokeRect(x, y, this.width, this.height);
+						}
+
+						// Set stroke on the inside of the box
+						else if (this.strokePosition === "inside") {
+							canvas.strokeRect(x + this.strokeWidth / 2, y + this.strokeWidth / 2, this.width - this.strokeWidth, this.height - this.strokeWidth);
+						}
+
 					}
-					
-					// Set stroke on the edge of the box (half of the stroke outside, half inside)
-					else if (this.strokePosition === "center") {
-						canvas.strokeRect(x, y, this.width, this.height);
+
+					// Do clip
+					if (this.clipChildren) {
+						canvas.clip();
 					}
-					
-					// Set stroke on the inside of the box
-					else if (this.strokePosition === "inside") {
-						canvas.strokeRect(x + this.strokeWidth / 2, y + this.strokeWidth / 2, this.width - this.strokeWidth, this.height - this.strokeWidth);
+
+				} else {
+
+					canvas.beginPath();
+					drawBorderRadiusRect(canvas, x, y, this.width, this.height, borderRadius);
+					canvas.closePath();
+
+					// Do fill if a color is specified
+					if (this.fill !== "") {
+						canvas.fillStyle = this.fill;
+						canvas.fill();
 					}
-				}
 
-				canvas.closePath();
+					var strokeSimulatedUsingPath = false;
+
+					// Do color if stroke width is specified
+					if (this.strokeWidth > 0) {
+
+						if (this.strokePosition === 'center') {
+
+							// Set styles
+							canvas.lineWidth = this.strokeWidth;
+							canvas.strokeStyle = this.strokeColor;
+
+							canvas.stroke();
+
+						} else {
+
+							// Simulate stroke using a path
+
+							canvas.beginPath();
+
+							if (this.strokePosition === 'inside') {
+								drawBorderRadiusRect(canvas, x, y, this.width, this.height, borderRadius);
+								drawBorderRadiusRect(canvas, x + this.strokeWidth, y + this.strokeWidth, this.width - this.strokeWidth * 2, this.height - this.strokeWidth * 2, addAmountToBorderRadius(borderRadius, -this.strokeWidth));
+							} else if (this.strokePosition === 'outside') {
+								drawBorderRadiusRect(canvas, x - this.strokeWidth, y - this.strokeWidth, this.width + this.strokeWidth * 2, this.height + this.strokeWidth * 2, addAmountToBorderRadius(borderRadius, this.strokeWidth));
+								drawBorderRadiusRect(canvas, x, y, this.width, this.height, borderRadius);
+							}
+
+							canvas.closePath();
+
+							canvas.fillStyle = this.strokeColor;
+							canvas.fill('evenodd');
```

Comment:
It would be nice if we didn't need to use `'evenodd'` here, since that's not supported in IE9 and IE10. Even though the usage numbers of IE9 and IE10 might be getting low now, the library still states that as the browser support level.

You should be able to simulate it by drawing the second shape (line 176) counter-clockwise, so maybe you can give that a try! :)

Reply:
Done! :) drawing the second path counter-clockwise was my first approach, but it didn't work because I was doing something wrong at other point of the code. I have done the change again, and it's working perfectly, as you predicted. Thanks!

I just realize I wrote "clockwise" in the commit message, but I wanted to write "counter clockwise". There is no way to change that message, there is?

Reply:
You can do `git commit --amend` and change the message of the last commit :)

Reply:
Looks like something got messed up with the commits. You can do `git rebase develop -i` and remove the last three commits (at the top) by just removing the lines in there, then replace `pick` with `r` for the commit you want to change the commit message of. Then you force push with `git push -f` since you need to overwrite what's on your branch on your fork currently.

Reply:
I don't know what I did yesterday. I normally use git inside of IntelliJ IDEA, and I think that was the problem. I was waiting for today in order to fix it. Thank you for help me with those commands, but I have encountered a problem. When I do `git rebase develop -i`, this is what I see:

```
noop

# Rebase 09e5ded..09e5ded onto 09e5ded (1 command(s))
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
# d, drop = remove commit
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out
```

No commits at all, only a `noop` line. Would you help me, please? I always get stuck when I try to use git beyond commit and push :sweat_smile: 

Thank you again!

Reply:
Oh sorry, I didn’t see you were working on `develop` on your fork, I assumed you were on another branch. I’m not sure how you cloned and set up remotes, but you need to rebase on top of `develop` from `koggdal/ocanvas`. You can also do `git log` and copy the commit hash of the commit right before your first commit, then `git rebase <hash> -i`.

Reply:
So the right way to go is to create a new branch when programming new features, right? I tried to rebase, put it's giving me some errors. I'm not a git expert, so I don't know how to solve them. What do you think if I delete my fork, then I do another fork, put the changes in it, and create another PR?

Thank you very much!

Reply:
Yeah normally one would check out a new branch for each thing. For single contributions like this maybe it doesn't matter much, but when working a lot on a project you might have multiple PRs open at the same time, and then you definitely need multiple branches. I tend to just make it simple for me so I don't have to think, and always create new branches for whatever I'm working on :) It also makes it easy to rebase in case new changes come into the project from other people.

Not sure what errors you got (works fine when I try running `git rebase origin/develop -i` (origin being my repo `koggdal/ocanvas`), deleting the bottom two lines and rewording the remaining commit that you wanted to change the message of by replacing `pick` by `r`). If that doesn't work for whatever reason you can also do:

``` shell
# assuming you have two remotes (origin and koggdal) and your local branch with your work is `develop`

# create a new temp branch based on latest `develop` branch, with the right commits
git checkout koggdal/develop
git checkout -b tempbranch
git cherry-pick 1924832607c8890c5f92222dd47250b17e44e3de
git cherry-pick f33c065abf33465d9ac23797e622c59362a99411
git cherry-pick c22ace74bda8b7cb2d571031a2448acb089b81e2

# reword the commit message
git commit --amend

# put it back on the `develop` branch and push it up to this PR
git branch -D develop
git checkout -b develop
git branch -D tempbranch
git push -u origin develop
```

Of course you can also just kill the whole fork and start over, but it's a good learning opportunity to get into git a bit, and not lose the commits :)

Reply:
Hi!

I only had one remote (origin, refering to my fork), so I added a new remote using:

```
$ git remote add koggdal https://github.com/koggdal/ocanvas.git
```

Then I tried to do the checkout, but I'm getting the next error:

```
$ git remote
koggdal
origin
$ git checkout koggdal/develop
error: pathspec 'koggdal/develop' did not match any file(s) known to git.
```

Sorry for the inconveniences, I'm trying to understand what I'm doing, but I don't get it :sweat_smile:

Reply:
No inconvenience, I'm happy to help :)

Did you do `git fetch koggdal` after adding the remote?

Reply:
And oh, I guess if you didn't have my remote added before, that was probably the reason the rebase didn't work properly? That would have been a `git rebase koggdal/develop -i` with the remote named `koggdal`. Names might be confusing because it's all just conventions, which different people do differently.. :) Personally when I clone I tend to clone the main project and have that as the `origin` remote, then add my own fork as `koggdal`, but that's of course not how everybody does it :)

Reply:
In which step did you get a conflict? If you follow those steps it should work. When you do `git checkout koggdal/develop`, can you verify that the last commit then is `63d5eb3a162e62ad5f8e36ad030623bc36c22eb5`? Applying the three mentioned commits after that (in that order) should be fine without conflicts. In the last step with `git push` I forgot to add that you need to add `-f` for force (since you're overwriting whatever you pushed to your fork before), so that would be `git push -f -u origin develop`.

I don't know the tutorial landscape for git really, sorry.. but google and you'll probably find something good :)

Reply:
My general good tips for working with git:

* Clone main repo, then add your fork as a remote. When working with multiple people it's easy and nice to think of different remotes (`origin` is always sort of what it says, the central origin of it all, `koggdal` is my own fork, `nestorrente` is your fork).

* **Always** start by checking out a new branch. This will save you from other hard things later. If you know that the main branches (`master` and `develop` in the case of oCanvas) never contain your own changes that haven't been merged in yet, you can always rely on these branches for things like rebasing.
  * For example, if this PR had been done from a custom branch, you could easily have done `git rebase develop -i` and fixed it all in a single rebase step, since `develop` then refers to the `develop` branch of `origin` (`koggdal/ocanvas` repo). If other people pushed changes to the main repo and they conflict with your changes, or just that you want those changes before yours, you can simply do:

    ```
    git checkout develop
    git pull
    git checkout -
    git rebase develop
    ```

  * If you need to pause your work and do something else, that is much easier if you've been working on a custom branch.

    ```
    git commit -a -n -m "wip"
    git checkout develop
    git checkout -b my-new-thing
    ```

  * If you have multiple PRs open, you need multiple branches since each PR points to a branch. Maybe you didn't intend on having multiple PRs, but maybe the first one dragged out and took a long time and you started on a secondary thing while waiting for something. Then it's much easier to start working on the second thing if `develop` is clean with the latest state without any changes from open PRs and you can just start a new branch from there.

* Use interactive rebase a lot as your way of working when you have multiple commits. `git rebase develop -i` will give you a nice overview of the commits you're working on. It's also really useful for many things. Maybe you want to change the message of the first commit. Maybe you want to change the order of two commits. Or delete a commit entirely. Or squash some commits together into one commit.

* Write [good commit messages](https://chris.beams.io/posts/git-commit/). This helps yourself in understanding what the commit is doing, and you might realize you missed something or could do something in a better way, just by trying to explain it. It also helps reviewers and future developers (maybe yourself!) trying to understand some code a few years later. It's also really important to describe _why_ the something was done (in the cases where it makes sense), to help future developers tracing back the reason for code being in a certain way.

* Use `git add -p` a lot. This will make sure that you know what you're committing. It also allows you to make a bunch of changes to explore how you want to build something maybe, then make multiple sensible commits from all those changes.

This is most of my git workflow. Knowing my remotes, knowing which branches I have and being certain of what content is on which branches, making good commits and using rebase to help me do that and always be up-to-date. Hope you find some of that useful! :) And go google for some other git tutorials for the specific concepts you want more knowledge of :)

Reply:
Hi!

Sorry, I was busy and I let this unanswered. I'm not able to fix my develop branch, so I deleted the previous fork and created a new one. I'll close this PR and created a new one right now.

Despite this, thank you very much! I'll follow your recommendations the next time :)